compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -398,6 +398,10 @@ impl server::FreeFunctions for Rustc<'_> {
     fn track_env_var(&mut self, var: &str, value: Option<&str>) {
         self.sess.env_depinfo.borrow_mut().insert((Symbol::intern(var), value.map(Symbol::intern)));
     }
+
+    fn track_path(&mut self, path: &str) {
+        self.sess.file_depinfo.borrow_mut().insert(Symbol::intern(path));
```

Comment:
Should this use `SourceMap::load_file` instead? I know this reads the file but this both checks if the file exists and handles `--remap-path-prefix`. An additional benefit is that we can use `&Path`s rather than UTF-8 `&str` paths and we can leverage the existing "read `source_map().files()` to Makefile-compatible .d file` logic below.

@petrochenkov Do we need something new like `SourceMap::track_file` for this use case?

Reply:
If we agree on this, there needs some mitigation on using `track_path!` or `tracked::fs::path` on a regular source file that is already tracked a source file.

Reply:
>Should this use `SourceMap::load_file` instead?

I don't think so.
Dep info is filled with `unmapped_path`s, and the `is_real_file` check is supposed to filter away things like `<macros>`, not non-existing files.
Non-existing files or env vars in depinfo are a normal thing - think about reading an optional config (we need to rebuild when it appears).

Besides that `load_file` doesn't do anything except for actually loading the file, which we certainly should not do in the tracking API.

>An additional benefit is that we can use `&Path`s rather than UTF-8 `&str`

Actually, feel free to use `Path`s in the tracking API and keep them in `sess.file_depinfo` if you can pass them through the proc macro bridge.
During writing to depinfo files they will be corrupted anyway though if they contain non-UTF-8 stuff, this is is a pre-existing issue which applies to paths in the source map as well.

>If we agree on this, there needs some mitigation on using `track_path!` or `tracked::fs::path` on a regular source file that is already tracked a source file.

Currently we don't need any such mitigations, depinfo can contain duplicate paths.